File: src/cpp/multithreading/merge_sort.cpp
/**
 * Parallel Merge Sort
 *
 * This example demonstrates divide-and-conquer parallelism using merge sort.
 * The algorithm splits work recursively, with each thread handling a subtree.
 *
 * Key concepts:
 * - Divide-and-conquer parallelism: split problem, solve in parallel, combine
 * - Depth-limited parallelism: stop spawning threads beyond a threshold
 * - Thread overhead: too many threads can hurt performance
 * - Compare parallel vs sequential to see actual speedup
 *
 * The depth limit prevents creating too many threads. Beyond a certain depth,
 * sequential execution is more efficient due to thread creation overhead.
 */

#include <algorithm>
#include <chrono>
#include <cmath>
#include <iostream>
#include <thread>
#include <vector>

// Merge two sorted subarrays: [left..mid] and [mid+1..right]
void merge(std::vector<int>& arr, int left, int mid, int right) {
    std::vector<int> temp(right - left + 1);
    int i = left, j = mid + 1, k = 0;

    while (i <= mid && j <= right) {
        if (arr[i] <= arr[j]) {
            temp[k++] = arr[i++];
        } else {
            temp[k++] = arr[j++];
        }
    }

    while (i <= mid) temp[k++] = arr[i++];
    while (j <= right) temp[k++] = arr[j++];

    std::copy(temp.begin(), temp.end(), arr.begin() + left);
}

// Sequential merge sort
void sequential_sort(std::vector<int>& arr, int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        sequential_sort(arr, left, mid);
        sequential_sort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

// Parallel merge sort with depth limiting
void parallel_sort(std::vector<int>& arr, int left, int right, int depth = 0) {
    // Stop spawning threads beyond log2(num_cores) depth
    // Handle case where hardware_concurrency() returns 0
    unsigned int cores = std::thread::hardware_concurrency();
    if (cores == 0) cores = 2;  // Default to 2 if unknown
    int max_depth = static_cast<int>(std::log2(cores));

    if (left >= right) return;

    int mid = left + (right - left) / 2;

    if (depth < max_depth) {
        // Spawn thread for left half, do right half ourselves
        std::thread left_thread(parallel_sort, std::ref(arr),
                                left, mid, depth + 1);
        parallel_sort(arr, mid + 1, right, depth + 1);
        left_thread.join();
    } else {
        // Beyond depth limit, use sequential sort
        sequential_sort(arr, left, mid);
        sequential_sort(arr, mid + 1, right);
    }

    merge(arr, left, mid, right);
}

void benchmark(int size, int runs) {
    // Initialize array with random values
    std::vector<int> original(size);
    for (int i = 0; i < size; ++i) {
        original[i] = rand();
    }

    using namespace std::chrono;

    // Benchmark sequential
    std::cout << "Evaluating Sequential Merge Sort...\n";
    duration<double> seq_time(0);
    std::vector<int> seq_arr = original;
    sequential_sort(seq_arr, 0, size - 1);

    for (int r = 0; r < runs; ++r) {
        std::vector<int> arr = original;
        auto start = high_resolution_clock::now();
        sequential_sort(arr, 0, size - 1);
        seq_time += high_resolution_clock::now() - start;
    }
    seq_time /= runs;

    // Benchmark parallel
    std::cout << "Evaluating Parallel Merge Sort...\n";
    duration<double> par_time(0);
    std::vector<int> par_arr = original;
    parallel_sort(par_arr, 0, size - 1);

    for (int r = 0; r < runs; ++r) {
        std::vector<int> arr = original;
        auto start = high_resolution_clock::now();
        parallel_sort(arr, 0, size - 1);
        par_time += high_resolution_clock::now() - start;
    }
    par_time /= runs;

    // Verify results
    if (seq_arr != par_arr) {
        std::cerr << "ERROR: Results don't match!\n";
        return;
    }

    std::cout << "\nResults (" << size << " elements):\n";
    std::cout << "  Sequential: " << seq_time.count() * 1000 << " ms\n";
    std::cout << "  Parallel:   " << par_time.count() * 1000 << " ms\n";
    std::cout << "  Speedup:    " << seq_time / par_time << "x\n";
}

int main() {
    benchmark(100000, 5);
    return 0;
}

File: src/python/multithreading/merge_sort.py
"""
Threaded Merge Sort Example

This script demonstrates parallelizing the merge sort algorithm using threads.
It compares the performance of single-threaded vs multi-threaded implementations.

Key Concepts:
- Divide-and-conquer algorithms can benefit from parallelism
- ThreadPoolExecutor simplifies thread management
- Limiting recursion depth prevents thread explosion
- For CPU-bound tasks, threading may not help due to Python's GIL

Implementation Details:
- max_depth limits how deep we create new threads (prevents overhead)
- ThreadPoolExecutor manages thread lifecycle automatically
- Beyond max_depth, we fall back to regular recursion

Note: Due to Python's GIL, CPU-bound parallel merge sort may not be faster
with threads. For true parallelism with CPU-bound tasks, use multiprocessing.
This example still demonstrates the pattern for educational purposes.
"""

import random
import time
from concurrent.futures import ThreadPoolExecutor


def merge(left, right):
    """Merge two sorted lists into one sorted list."""
    result = []
    left_index = 0
    right_index = 0

    while left_index < len(left) and right_index < len(right):
        if left[left_index] < right[right_index]:
            result.append(left[left_index])
            left_index += 1
        else:
            result.append(right[right_index])
            right_index += 1

    # Append remaining elements
    result.extend(left[left_index:])
    result.extend(right[right_index:])

    return result


def merge_sort(arr):
    """Standard single-threaded merge sort."""
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return merge(left, right)


def threaded_merge_sort(arr, depth=0, max_depth=2):
    """
    Multi-threaded merge sort with configurable parallelism depth.

    Args:
        arr: List to sort
        depth: Current recursion depth (for limiting thread creation)
        max_depth: Maximum depth to create new threads
    """
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = arr[:mid]
    right = arr[mid:]

    if depth < max_depth:
        # Parallelize at shallow depths
        with ThreadPoolExecutor(max_workers=2) as executor:
            left_future = executor.submit(
                threaded_merge_sort, left, depth + 1, max_depth
            )
            right_future = executor.submit(
                threaded_merge_sort, right, depth + 1, max_depth
            )
            left = left_future.result()
            right = right_future.result()
    else:
        # Fall back to sequential sorting at deeper levels
        left = threaded_merge_sort(left, depth + 1, max_depth)
        right = threaded_merge_sort(right, depth + 1, max_depth)

    return merge(left, right)


def main():
    arr = [random.randint(1, 10000) for _ in range(10000)]

    # Benchmark non-threaded version
    start_time = time.time()
    sorted_arr = merge_sort(arr.copy())
    end_time = time.time()
    print(f"Non-threaded merge sort time: {end_time - start_time:.5f} seconds")

    # Benchmark threaded version
    start_time = time.time()
    sorted_arr_threaded = threaded_merge_sort(arr.copy())
    end_time = time.time()
    print(f"Threaded merge sort time: {end_time - start_time:.5f} seconds")

    # Verify both produce the same result
    assert sorted_arr == sorted_arr_threaded, "Results don't match!"
    print("Both methods produced identical sorted results.")


if __name__ == "__main__":
    main()
